compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_linalg_generic_on_tensors.mlir

```diff
@@ -0,0 +1,229 @@
+//      CHECK: ins(%[[IN]] : tensor<8x8x32x32x8192xf32>) outs(%[[OUT]] : tensor<8x8x32x32x8192xf32>)
+//      CHECK:  tensor.expand_shape %[[RES]] {{\[}}[0, 1], [2], [3], [4], [5, 6]] : tensor<8x8x32x32x8192xf32> into tensor<2x4x8x32x32x64x128xf32>
+
+// -----
+!type = tensor<2x4x8x16x32x64x128x256xf32>
+util.global private @"__transpose_10_input" {noinline} = dense<1.0> : !type
+
+func.func @collapse() -> !type {
+  %cst = arith.constant 0.000000e+00 : f32
+  %c0 = arith.constant 0 : index
+  %input_ptr = util.global.address @"__transpose_10_input" : !util.ptr<!type>
+  %input = util.global.load.indirect %input_ptr : !util.ptr<!type> -> !type
+  %output = tensor.empty() : !type
+  
+  // Can collapse (d0, d1) and (d3, d4, d5, d6, d7)
+  %result = linalg.generic { indexing_maps = [ 
+          affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d2, d3, d4, d5, d6, d7)>, 
+          affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d2, d3, d4, d5, d6, d7)>],
+          iterator_types = ["parallel", "parallel", "reduction", "parallel", "parallel", "parallel", "parallel", "parallel"]
+        }
+  ins(%input : !type) outs(%output : !type) {
+  ^bb0(%arg1: f32, %arg2: f32):
+    linalg.yield %arg1 : f32
+  } -> !type
+  return %result: !type
+
+}
+
+//      CHECK: %[[IN:.+]] = tensor.collapse_shape %[[INPUT:.+]] {{\[}}[0, 1], [2], [3, 4, 5, 6, 7]] : tensor<2x4x8x16x32x64x128x256xf32> into tensor<8x8x1073741824xf32>
+//      CHECK: %[[OUT:.+]] = tensor.collapse_shape %[[OUTPUT:.+]] {{\[}}[0, 1], [2], [3, 4, 5, 6, 7]] : tensor<2x4x8x16x32x64x128x256xf32> into tensor<8x8x1073741824xf32>
+//      CHECK: %[[RES:.+]] = flow.dispatch.region
+//      CHECK: linalg.generic {indexing_maps = [#[[MAP:.+]], #[[MAP:.+]]], iterator_types = ["parallel", "reduction", "parallel"]}
+//      CHECK: ins(%[[IN]] : tensor<8x8x1073741824xf32>) outs(%[[OUT]] : tensor<8x8x1073741824xf32>)
+//      CHECK:  tensor.expand_shape %[[RES]] {{\[}}[0, 1], [2], [3, 4, 5, 6, 7]] : tensor<8x8x1073741824xf32> into tensor<2x4x8x16x32x64x128x256xf32>
+
+// -----
+
+!type = tensor<2x4x8x16x64x64x128x256xf32>
+util.global private @"__transpose_10_input" {noinline} = dense<1.0> : !type
+func.func @collapse() -> !type {
+  %cst = arith.constant 0.000000e+00 : f32
+  %c0 = arith.constant 0 : index
+  %input_ptr = util.global.address @"__transpose_10_input" : !util.ptr<!type>
+  %input = util.global.load.indirect %input_ptr : !util.ptr<!type> -> !type
+  %output = tensor.empty() : !type
+  
+  // Can collapse (d0, d1) and (d6, d7)
+  %result = linalg.generic { indexing_maps = [ 
+          affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d2, d3, d4, d5, d6, d7)>, 
+          affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d2, d3, d5, d4, d6, d7)>],
+          iterator_types = ["parallel", "parallel", "reduction", "parallel", "parallel", "parallel", "parallel", "parallel"]
```

Comment:
in general the reduction dim wouldn't be there in the destination (otherwise the dim doesn't really need to be a reduction). Would this work for those maps:
```
affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d2, d3, d4, d5, d6, d7)>, 
affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d3, d5, d4, d6, d7)>],
```

Reply:
Yes, first two and last two loops are collapsed here.
I added this case as a test.